### PR TITLE
live: add JS, rework dots

### PR DESCRIPTION
**Pull request by kevboh**

### CODE REVIEW

File: live/js.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
+package live
+
+import (
+	"encoding/json"
+	"strings"
+)
+
+// DefaultTransitionDuration_ms is the default duration, in milliseconds, of live.JS transitions.
+const DefaultTransitionDuration_ms = int32(200)
```
Comment (josharian) [inline live/js.go:9]: Can we instead do `DefaultTransitionDuration = 200 * time.Millisecond` and then call `.Milliseconds()` on it when needed?

File: live/js.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+// JS provides a way to precompose simple client-side DOM changes that don't require a round-trip to the server.
+// Create a JS struct, call its methods to build up a command, and then render it as the value of a phx-* attribute.
+type JS struct {
+	ops []*op
+}
+
+// Hide hides elements.
+// When the action is triggered on the client, phx:hide-start is dispatched to the hidden elements. After the time specified by :time, phx:hide-end is dispatched.
```
Comment (josharian) [inline live/js.go:23]: s/:time/opts.Time/

File: live/js.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	EndClass string
+}
+
+type HideOpts struct {
+	// To is the DOM selector of the element to hide, or empty to target the interacted element.
+	To string
+	// Transition to apply, if any.
+	Transition *Transition
+	// Time is the duration in milliseconds of the transition, if present; defaults to DefaultTransitionDuration_ms if 0.
+	Time int32
```
Comment (josharian) [inline live/js.go:83]: Can we make this a time.Duration instead? If not, I understand. But it would be prettier :)

File: live/js.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	// Display is the CSS display value to set when showing; defaults to "block".
+	Display string
+}
+
+func (o *ShowOpts) args() map[string]any {
+	var to *string
+	time := DefaultTransitionDuration_ms
+	transition := &Transition{}
```
Comment (josharian) [inline live/js.go:120]: `var transition Transition` ?

File: live/js.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	// Time is the duration in milliseconds of the relevant transition, if present; defaults to DefaultTransitionDuration_ms if 0.
+	Time int32
+	// Display is the CSS display value to set when showing; defaults to "block".
+	Display string
+}
+
+func (o *ToggleOpts) args() map[string]any {
+	var to *string
+	time := DefaultTransitionDuration_ms
+	in := &Transition{}
```
Comment (josharian) [inline live/js.go:158]: `var in, out Transition`?

File: live/js.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+	var to *string
+	time := DefaultTransitionDuration_ms
+	transition := &Transition{}
+	display := "block"
+	if o != nil {
+		to = &o.To
+		if o.Time != 0 {
+			time = o.Time
+		}
+		if o.Transition != nil {
```
Comment (josharian) [inline live/js.go:127]: I don't think you need these if statements. When nil, You'll overwrite nil with nil.  no harm done.

File: live/js.go [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		"to":      to,
+		"ins":     in.phx(),
+		"outs":    out.phx(),
+		"display": display,
+	}
+}
+
+func phxClasses(class string) []string {
+	if class == "" {
+		return []string{}
```
Comment (josharian) [inline live/js.go:187]: Does it need to be an empty slice, or can it be nil?

File: readme.md [hunk 8]
```diff
@@ -156,6 +158,12 @@ r.HandleFunc("/user/{user_id}", func(w http.ResponseWriter, r *http.Request) {
 > **Note**  
 > When you patch a view in GoLive, we first give you an opportunity to re-handle the “request,” parsing it as needed, before calling `HandleParams`. In Phoenix terms, path params are handled different from URL query params: path params are parsed out at the muxer layer, URL query params in the more traditional `HandleParams` callback. This is a consequence of our decision to let you bring your own muxer, but may be unexpected for those familiar with Phoenix.
 
+## live.JS
+
+GoLive includes a struct, `JS`, that provides API to precompose client-side commands that do not require a roundtrip to the server, [much like Phoenix.LiveView.JS](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveView.JS.html) does. This is useful for doing light DOM manipulation without writing JavaScript, and is a feature of the `phoenix_live_view` JavaScript client protocol.
+
+Currently only `Hide`, `Show`, and `Toggle` are implemented; we plan to implement the remaining utility commands soon.
```
Comment (josharian) [inline readme.md:165]: want to add scare quotes around soon? :)

**Review by josharian — APPROVED** [review]

File: live/js.go [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+// JS provides a way to precompose simple client-side DOM changes that don't require a round-trip to the server.
+// Create a JS struct, call its methods to build up a command, and then render it as the value of a phx-* attribute.
+type JS struct {
+	ops []*op
+}
+
+// Hide hides elements.
+// When the action is triggered on the client, phx:hide-start is dispatched to the hidden elements. After the time specified by :time, phx:hide-end is dispatched.
+func (js *JS) Hide(opts *HideOpts) {
```
Comment (floodfx) [inline live/js.go:24]: Any reason not to make these "Fluent Interface" style?
```go
func (js *JS) Show(opts *ShowOpts) *JS {
	js.add("show", opts)
	return js
}
```
Then you can chain a bunch if need be (particularly inside a template):
```go
js.Show(...).Push(...)
```

Comment (floodfx) [reply live/js.go:24]: Oh right.  Go templates don't make calling these very easy at all...  No need to add rn.

**Review by floodfx — APPROVED** [review]

File: live/js.go (hunk 9 above)
Comment (kevboh) [reply live/js.go:24]: @floodfx Yeah, the entire motivation to support methods is that at present `JS` isn't intended to be called inside a template at all—originally I had everything variadic to support it, but the optional params make it tough. Easier to wrap it in view methods and call those. The plan is to see what we/others do in terms of wrapping `JS` in funcmaps, then perhaps implement a good version of those wrappers.

File: live/js.go (hunk 2 above)
Comment (kevboh) [reply live/js.go:23]: This is what I get for copying straight from the hex docs!

File: live/js.go (hunk 6 above)
Comment (kevboh) [reply live/js.go:127]: We don't want nil, is the thing—`phx-` attributes expect a 3-tuple of empty js arrays if there's no transition (`[[],[],[]]`)

Comment (kevboh) [reply live/js.go:127]: I suppose I could do an equivalent check in `Transition.phx` instead, and return the empty state there if it's nil

Comment (kevboh) [reply live/js.go:127]: I'll do that.

File: live/js.go (hunk 5 above)
Comment (kevboh) [reply live/js.go:158]: One day I'll remember you can do this in Go

File: live/js.go (hunk 7 above)
Comment (kevboh) [reply live/js.go:187]: I note this above, but the phoenix js client expects an empty js array as representing "no transition class"

**Pull request merged**